src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -31,119 +34,166 @@ public LocaleService(LocaleDTO dto) {
-            return dispMap;
+            return cacheItem;
         }
-        dispMap = getSupportedLanguagesFromDS(locale, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
-        if (dispMap != null && dispMap.size() > 0) {
-            logger.debug("Find the displayNames for product [{}], version [{}], locale [{}].\n", dto.getProductID(), dto.getVersion(), locale);
-            formattingCacheService.addSupportedLanguages(dto, locale, dispMap);
-            logger.debug("DisplayNames is cached for product [{}], version [{}], locale [{}]!\n\n", dto.getProductID(), dto.getVersion(), locale);
-            return dispMap;
-        }
-        return dispMap;
+        cacheItem = new LocaleCacheItem();
+        getSupportedLanguagesFromDS(locale, cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+        formattingCacheService.addSupportedLanguages(dto, locale, cacheItem);
+        logger.debug("DisplayNames is cached for product [{}], version [{}], locale [{}]!\n\n", dto.getProductID(), dto.getVersion(), locale);
+        return cacheItem;
     }
 
 
-    private Map<String, String> getSupportedLanguagesFromDS(String locale, ListIterator<DataSourceEnum> msgSourceQueueIter) {
+    private void getSupportedLanguagesFromDS(String locale, LocaleCacheItem cacheItem, ListIterator<DataSourceEnum> msgSourceQueueIter) {
         Map<String, String> dispMap = new HashMap<String, String>();
         if (!msgSourceQueueIter.hasNext()) {
-            logger.error(ConstantsMsg.GET_LANGUAGES_FAILED_ALL);
-            return dispMap;
+            logger.error(ConstantsMsg.GET_LANGUAGES_FAILED_ALL, locale);
+            return;
         }
+        long timestampOld = cacheItem.getTimestamp();
         DataSourceEnum dataSource = (DataSourceEnum) msgSourceQueueIter.next();
-        dispMap = dataSource.createLocaleOpt(dto).getSupportedLanguages(locale);
-        if (dispMap == null || dispMap.isEmpty()) {
-            logger.debug(ConstantsMsg.GET_LANGUAGES_FAILED, dataSource.toString());
-            dispMap = getSupportedLanguagesFromDS(locale, msgSourceQueueIter);
+        dataSource.createLocaleOpt(dto).getSupportedLanguages(locale, cacheItem);
+        long timestampNew = cacheItem.getTimestamp();
+        if (timestampNew == timestampOld) {
+            logger.debug(ConstantsMsg.GET_LANGUAGES_FAILED, locale, dataSource.toString());
+        }
+        // Skip this block if timestamp is not 0 (which means cacheItem is in the cache) regardless if cacheItem is expired or not.
+        // Otherwise, try the next dataSource in the queue.
+        if (timestampNew == 0) {
+            getSupportedLanguagesFromDS(locale, cacheItem, msgSourceQueueIter);
         }
-        return dispMap;
+    }
+
+    private void populateRegionsCache(String locale, LocaleCacheItem cacheItem) {
+        Callable<LocaleCacheItem> callable = () -> {
+            try {
+
+                // Pass cacheItem to getMessages so that:
```

Comment:
Remove all these comments, they are for translations.

Reply:
or just change "getMessages" to "getRegionsFromDS". Better yet, put this explanation in a javadoc block.